### PR TITLE
Add PROBE_STACK to JSONParser::Walk (Fix #3900)

### LINKED ISSUE #3900: Stackoverflow in JSON.parse
If the reviver changes the result of JSON.parse, it can make infinite recursive calls, which trigger the stackoverflow.

This is a PoC code.
```
var o = {0:1,1:1};
JSON.parse('[0,0]', function () {
    this[1] = o;
})
```

### PR DESCRIPTION
**Pull request by xiaoyinl**
This PR fixes the infinite recursion bug in JSON.parse.

Fixes #3900 

/cc @DaramG @MikeHolman

### CODE REVIEW

**Comment by Penguinwizzard** [comment]
Could you add a unit test for this?

File: test/JSON/stackoverflow.js [hunk 1]
```diff
@@ -0,0 +1,17 @@
...
+{
+    // Issue Microsoft/ChakraCore#3900
+    var o = { 0: 1, 1: 1 };
+    JSON.parse('[0,0]', function () {
+        this[1] = o;
+    });
+}
+catch (ex)
+{
+    writeLine(ex.name + ": " + ex.message);
```
Comment (obastemur) [inline test/JSON/stackoverflow.js]: nit: `console.log` instead

File: test/JSON/stackoverflow.baseline [hunk 2]
```diff
@@ -0,0 +1 @@
+Error: Out of stack space
```
Comment (obastemur) [inline test/JSON/stackoverflow.baseline]: nit: could be nice to ditch `baseline file` and compare output and `console.log('PASS')` ?

**Comment by obastemur** [comment]
LGTM

**Comment by xiaoyinl** [comment]
@obastemur All tests passed; both comments are addressed. Thus I squashed the commits.

**Pull request merged**